### PR TITLE
RavenDB-21651 CSRF protection in Studio

### PR DESCRIPTION
**Pull request by ml054**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21651 

### Additional description

CSRF Protection for RavenDB server

### Type of change

- New feature

### How risky is the change?

- High

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works
- It has been verified by manual testing

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Web/RequestHandler.cs [hunk 1]
```diff
@@ -40,6 +40,8 @@ public abstract class RequestHandler
 
         public const string PageSizeParameter = "pageSize";
 
+        public static readonly List<string> SafeCsrfMethods = new(new[] { "HEAD", "OPTIONS", "TRACE" });
```
Comment (ppekrol) [inline src/Raven.Server/Web/RequestHandler.cs]: HashSet + do:
`new() { "Val", "Val2", "Val3" }`

File: src/Raven.Server/Web/RequestHandler.cs [hunk 2]
```diff
@@ -687,6 +689,61 @@ private static void ThrowInvalidAuthStatus(RavenServer.AuthenticationStatus? sta
             throw new ArgumentOutOfRangeException("Unknown authentication status: " + status);
         }
 
+        public static bool CheckCSRF(HttpContext httpContext, ServerStore serverStore)
+        {
+            if (serverStore.Configuration.Security.EnableCsrfFilter == false)
+                return true;
+            
+            var host = httpContext.Request.Host;
+            if (string.IsNullOrWhiteSpace(host.ToString()))
```
Comment (ppekrol) [inline src/Raven.Server/Web/RequestHandler.cs]: can we avoid ToString?

File: src/Raven.Server/Web/RequestHandler.cs [hunk 3]
```diff
@@ -687,6 +689,61 @@ private static void ThrowInvalidAuthStatus(RavenServer.AuthenticationStatus? sta
...
+                return false;
+            
+            var requestedOrigin = httpContext.Request.Headers["Origin"];
```
Comment (ppekrol) [inline src/Raven.Server/Web/RequestHandler.cs]: move Origin to Constants

File: src/Raven.Server/Web/RequestHandler.cs (hunk 2 above)
Comment (ppekrol) [reply src/Raven.Server/Web/RequestHandler.cs]: In what case this can be null?

File: src/Raven.Server/Web/RequestHandler.cs [hunk 4]
```diff
@@ -687,6 +689,61 @@ private static void ThrowInvalidAuthStatus(RavenServer.AuthenticationStatus? sta
...
+
+            if (SafeCsrfMethods.Contains(httpContext.Request.Method))
+                return true;
+            
+            var origin = requestedOrigin[0];
+            var originHost = new Uri(origin).Host;
+            var originAuthority = new Uri(origin).Authority;
+            
+            // for hostname matching we validate both hostname and port
+            var hostMatches = host.ToString() == originAuthority;
```
Comment (ppekrol) [inline src/Raven.Server/Web/RequestHandler.cs:722]: again ToString

File: src/Raven.Server/Web/RequestHandler.cs [hunk 5]
```diff
@@ -687,6 +689,61 @@ private static void ThrowInvalidAuthStatus(RavenServer.AuthenticationStatus? sta
...
+                return true;
+            
+            // for requests with-in cluster we value both hostname and port
+            var requestWithinCluster = IsOriginAllowed(origin, serverStore);
+            if (requestWithinCluster)
+                return true;
+            
+            // for trusted origins we match hostname only, port is ignored
+            var trustedOrigins = serverStore.Configuration.Security.CsrfTrustedOrigins ?? Array.Empty<string>();
+            foreach (var o in trustedOrigins)
```
Comment (ppekrol) [inline src/Raven.Server/Web/RequestHandler.cs]: check the count before doing foreach

File: src/Raven.Server/Web/RequestHandler.cs [hunk 6]
```diff
@@ -687,6 +689,61 @@ private static void ThrowInvalidAuthStatus(RavenServer.AuthenticationStatus? sta
...
+            {
+                if (originHost == o)
+                    return true;
+            }
+
+            // for additional origin headers we match hostname only, port is ignored
+            var additionalHeaders = serverStore.Configuration.Security.CsrfAdditionalOriginHeaders ?? Array.Empty<string>();
+            foreach (string additionalHeader in additionalHeaders)
```
Comment (ppekrol) [inline src/Raven.Server/Web/RequestHandler.cs]: check the count before doing foreach

File: src/Raven.Server/Web/RequestHandler.cs [hunk 7]
```diff
@@ -687,6 +689,61 @@ private static void ThrowInvalidAuthStatus(RavenServer.AuthenticationStatus? sta
...
+            {
+                if (httpContext.Request.Headers.TryGetValue(additionalHeader, out var headerValue) == false)
+                    continue;
+
+                var stringHeader = headerValue.ToString();
```
Comment (ppekrol) [inline src/Raven.Server/Web/RequestHandler.cs]: can we avoid doing that?

File: src/Raven.Server/Web/RequestHandler.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Server/Web/RequestHandler.cs]: internal

File: src/Raven.Server/Web/RequestHandler.cs [hunk 8]
```diff
@@ -687,6 +689,61 @@ private static void ThrowInvalidAuthStatus(RavenServer.AuthenticationStatus? sta
...
+            
+            if (requestedOrigin.Count == 0 || requestedOrigin[0] == null)
+                return true;
+
+            if (SafeCsrfMethods.Contains(httpContext.Request.Method))
+                return true;
+            
+            var origin = requestedOrigin[0];
+            var originHost = new Uri(origin).Host;
+            var originAuthority = new Uri(origin).Authority;
```
Comment (ppekrol) [inline src/Raven.Server/Web/RequestHandler.cs]: double URI parsing here, should we cache it?

File: src/Raven.Server/Web/RequestHandler.cs (hunk 1 above)
Comment (ml054) [reply src/Raven.Server/Web/RequestHandler.cs]: done

File: src/Raven.Server/Web/RequestHandler.cs (hunk 2 above)
Comment (ml054) [reply src/Raven.Server/Web/RequestHandler.cs]: done

null -> "should never happen"

File: src/Raven.Server/Web/RequestHandler.cs (hunk 3 above)
Comment (ml054) [reply src/Raven.Server/Web/RequestHandler.cs]: done

File: src/Raven.Server/Web/RequestHandler.cs (hunk 8 above)
Comment (ml054) [reply src/Raven.Server/Web/RequestHandler.cs]: done

File: src/Raven.Server/Web/RequestHandler.cs (hunk 4 above)
Comment (ml054) [reply src/Raven.Server/Web/RequestHandler.cs:722]: done

File: src/Raven.Server/Web/RequestHandler.cs (hunk 5 above)
Comment (ml054) [reply src/Raven.Server/Web/RequestHandler.cs]: done

File: src/Raven.Server/Web/RequestHandler.cs (hunk 6 above)
Comment (ml054) [reply src/Raven.Server/Web/RequestHandler.cs]: done

File: src/Raven.Server/Web/RequestHandler.cs (hunk 7 above)
Comment (ml054) [reply src/Raven.Server/Web/RequestHandler.cs]: what do you suggest?

File: src/Raven.Server/Web/RequestHandler.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Server/Web/RequestHandler.cs]:
```csharp
    public static readonly HashSet<string> SafeCsrfMethods = new()
    {
        HttpMethod.Head.Method,
        HttpMethod.Options.Method,
        HttpMethod.Trace.Method
    };
```

File: src/Raven.Server/Web/RequestHandler.cs [hunk 9]
```diff
@@ -687,6 +689,69 @@ private static void ThrowInvalidAuthStatus(RavenServer.AuthenticationStatus? sta
             throw new ArgumentOutOfRangeException("Unknown authentication status: " + status);
         }
 
+        public static bool CheckCSRF(HttpContext httpContext, ServerStore serverStore)
+        {
+            if (serverStore.Configuration.Security.EnableCsrfFilter == false)
+                return true;
+            
+            var host = httpContext.Request.Host;
+            var hostAsString = host.ToString();
```
Comment (ppekrol) [inline src/Raven.Server/Web/RequestHandler.cs]: please ping me about this

File: src/Raven.Server/Web/RequestHandler.cs (hunk 1 above)
Comment (ml054) [reply src/Raven.Server/Web/RequestHandler.cs]: done

**Pull request merged**